lib/puppet/confine/boolean.rb

```diff
@@ -0,0 +1,50 @@
+require 'puppet/confine'
+
+# Common module for the Boolean confines. It currently
+# contains just enough code to implement PUP-9336.
+class Puppet::Confine
+  module Boolean
+    # Returns the passing value for the Boolean confine. This should
+    # just be the class name.
+    def passing_value
+      return @passing_value unless @passing_value.nil?
+
+      class_name = self.class.to_s.split('::').pop.downcase.to_sym
+
+      @passing_value = class_name == :true ? true : false
+    end
+
+    # The Boolean confines 'true' and 'false' let the user specify
+    # two types of values:
+    #     * A lambda for lazy evaluation. This would be something like
+    #         confine :true => lambda { true }
+    #
+    #     * A single Boolean value, or an array of Boolean values. This would
+    #     be something like
+    #         confine :true => true OR confine :true => [true, false, false, true]
+    #
+    # This override distinguishes between the two cases.
+    def values
+      # Note that Puppet::Confine's constructor ensures that @values
+      # will always be an array, even if a lambda's passed in. This is
+      # why we have the length == 1 check.
+      unless @values.length == 1 && @values.first.respond_to?(:call)
+        return @values
+      end
+
+      # We have a lambda. Here, we want to enforce "cache positive"
+      # behavior, which is to cache the result _if_ it evaluates to
+      # the passing value (i.e. the class name).
+
+      return @cached_value unless @cached_value.nil?
```

Comment:
Why does the confine implement the cache positive logic? IIRC the `Puppet::Confine.test` method implements the caching logic. By default it's cache positive, but if `Puppet[:always_retry_plugins]` is false, then we cache negative also. This is needed for puppetserver, because we're not waiting for confines/features to become available.

Reply:
https://github.com/puppetlabs/puppet/pull/7305#discussion_r242535196 talks about the why behind the cache-positive logic. One thing I was concerned about there was the fact that, without the caching, the upstart confine would be evaluated multiple times (specifically every-time provider suitability is checked). Since that confine's a shell call, that may slow things down.

Also, [Puppet::Confine.test](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/confine.rb#L23-L38) caches the confine classes, not the values themselves?